usr/share/rear/layout/save/default/600_snapshot_files.sh

```diff
@@ -10,17 +10,40 @@
 # we risk inconsistencies between the restored files and recreated layout,
 # or failures of UUID migration.
 
-local file obj
+local file obj final_file symlink_target
 
 # The patterns are relative to /, change directory there
 # so that the shell finds the files during pathname expansion
 pushd / >/dev/null
 # The variable expansion is deliberately not quoted in order to perform
 # pathname expansion on the variable value.
 for file in $FILES_TO_PATCH_PATTERNS ; do
-    if ! IsInArray "/$file" "${CHECK_CONFIG_FILES[@]}" ; then
-        CHECK_CONFIG_FILES+=( "/$file" )
+    final_file="/$file"
+    # Symlink handling: replace symlinks by targets (not strictly necessary, but consistent
+    # with 280_migrate_uuid_tags.sh), avoid dead symlinks, and symlinks to files
+    # on dynamic filesystem ( /proc etc.) - they are expected to change
+    # and validating their checksums has no sense
+    if ! IsInArray "$final_file" "${CHECK_CONFIG_FILES[@]}" ; then
+        if test -L "$final_file" ; then
+            if symlink_target="$( readlink -f "$final_file" )" ; then
```

Comment:
I know the `readlink -f "$restored_file"` is a copy
of the code in 280_migrate_uuid_tags.sh where
`git log --follow -p usr/share/rear/finalize/GNU/Linux/280_migrate_uuid_tags.sh`
shows that code originated at the generic origin commit therein
in the file finalize/GNU/Linux/15_migrate_disk_devices.sh

But as far as I see `readlink -f` does not work as intended:
```
# cd /tmp

# ln -s QQQ dead_symlink

# file dead_symlink
dead_symlink: broken symbolic link to QQQ

# readlink -f QQQ && echo Y || echo N
/tmp/QQQ
Y

# readlink -e QQQ && echo Y || echo N
N
```

Reply:
Together with additional code simplifications
(checks first to `continue` early in the `for` loop - see above)
I suggest this better code that uses `readlink -e` instead of `readlink -f`
(untested proposal):
```
for file in $FILES_TO_PATCH_PATTERNS ; do
    IsInArray "$file" "${CHECK_CONFIG_FILES[@]}" && continue
    # Symlink handling: replace symlinks by targets (not strictly necessary, but consistent
    # with 280_migrate_uuid_tags.sh), avoid dead symlinks, and symlinks to files
    # on dynamic filesystem ( /proc etc.) - they are expected to change
    # and validating their checksums has no sense
    final_file="/$file"
    if test -L "$final_file" ; then
        if ! symlink_target="$( readlink -e "$final_file" )" ; then
            LogPrint "Skip adding dead symlink $final_file to CHECK_CONFIG_FILES"
            continue
        fi
        # If the symlink target contains /proc/ /sys/ /dev/ or /run/ we skip it because then
        # the symlink target is considered to not be a restored file that needs to be patched
        # and thus we don't need to generate and check its hash either,
        # see https://github.com/rear/rear/pull/2795#issuecomment-1113105441
        # and https://github.com/rear/rear/pull/2047#issuecomment-464846777
        if echo $symlink_target | egrep -q '/proc/|/sys/|/dev/|/run/' ; then
            Log "Skip adding symlink $final_file target $symlink_target on /proc/ /sys/ /dev/ or /run/ to CHECK_CONFIG_FILES"
            continue
        fi
        Debug "Adding symlink $final_file target $symlink_target to CHECK_CONFIG_FILES"
        final_file="$symlink_target"
    fi
    CHECK_CONFIG_FILES+=( "$final_file" )
done
```

Reply:
Thanks for the suggestion, I will do this (but without the `IsInArray "$symlink_target" "${CHECK_CONFIG_FILES[@]}" && continue` and `final_file="$symlink_target"` parts, because I now believe one should not use the symlink target, see above)

Reply:
I think we should check both where a symlink points to
and also automatically check symlink target contents, see
https://github.com/rear/rear/pull/2795#discussion_r862712166

Reply:
By the way, I checked that `readlink` supports `-e` also in RHEL 6 (the oldest version that we support).

Reply:
I checked that `readlink -e` works same as I tested above in
https://github.com/rear/rear/pull/2795#discussion_r862667013
also in SLES10 SP4 which provides
```
# readlink --version
readlink (GNU coreutils) 5.93
```

Reply:
I like to keep this conversation open because it is in general
about better use `readlink -e` instead of `readlink -f` and
we have several other code places with `readlink -f`
which may need to be fixed.